File: src/llm_node/llm_node/llm_node.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from std_msgs.msg import String
from sensor_msgs.msg import Image
import openai
import os
import json
from typing import Optional


class LLMNode(Node):
    """
    ROS2 node that listens to ROS events and calls OpenAI API for LLM responses.
    """
    
    def __init__(self):
        super().__init__('llm_node')
        
        # Initialize OpenAI client
        self.openai_client = None
        self._setup_openai()
        
        # Create subscribers for different message types
        self.string_subscription = self.create_subscription(
            String,
            '/llm_input',
            self.string_callback,
            10
        )
        
        self.image_subscription = self.create_subscription(
            Image,
            '/camera/image_raw',
            self.image_callback,
            10
        )
        
        # Create a publisher for responses
        self.response_publisher = self.create_publisher(
            String,
            '/llm_response',
            10
        )
        
        self.get_logger().info('LLM Node initialized and listening for events...')
    
    def _setup_openai(self):
        """Setup OpenAI client with API key from environment variable."""
        api_key = os.getenv('OPENAI_API_KEY')
        if not api_key:
            self.get_logger().error('OPENAI_API_KEY environment variable not set!')
            return
        
        try:
            self.openai_client = openai.OpenAI(api_key=api_key)
            self.get_logger().info('OpenAI client initialized successfully')
        except Exception as e:
            self.get_logger().error(f'Failed to initialize OpenAI client: {e}')
    
    def string_callback(self, msg: String):
        """Handle string messages from ROS topic."""
        if not self.openai_client:
            self.get_logger().warn('OpenAI client not available, skipping request')
            return
        
        self.get_logger().info(f'Received string message: {msg.data}')
        
        try:
            response = self._call_openai_api(msg.data)
            if response:
                self.get_logger().info(f'LLM Response: {response}')
                
                # Publish response back to ROS
                response_msg = String()
                response_msg.data = response
                self.response_publisher.publish(response_msg)
        
        except Exception as e:
            self.get_logger().error(f'Error processing string message: {e}')
    
    def image_callback(self, msg: Image):
        """Handle image messages from ROS topic."""
        if not self.openai_client:
            self.get_logger().warn('OpenAI client not available, skipping request')
            return
        
        self.get_logger().info(f'Received image message: {msg.width}x{msg.height}')
        
        try:
            # Convert image data to base64 for OpenAI API
            import base64
            import io
            from PIL import Image as PILImage
            
            # Convert ROS Image to PIL Image
            pil_image = PILImage.frombytes(
                'RGB',
                (msg.width, msg.height),
                msg.data
            )
            
            # Convert to base64
            buffer = io.BytesIO()
            pil_image.save(buffer, format='JPEG')
            image_base64 = base64.b64encode(buffer.getvalue()).decode()
            
            # Call OpenAI with image
            response = self._call_openai_api_with_image(
                "Describe what you see in this image.",
                image_base64
            )
            
            if response:
                self.get_logger().info(f'LLM Response: {response}')
                
                # Publish response back to ROS
                response_msg = String()
                response_msg.data = response
                self.response_publisher.publish(response_msg)
        
        except Exception as e:
            self.get_logger().error(f'Error processing image message: {e}')
    
    def _call_openai_api(self, prompt: str) -> Optional[str]:
        """Call OpenAI API with text prompt."""
        try:
            response = self.openai_client.chat.completions.create(
                model="gpt-3.5-turbo",
                messages=[
                    {"role": "user", "content": prompt}
                ],
                max_tokens=150,
                temperature=0.7
            )
            
            return response.choices[0].message.content.strip()
        
        except Exception as e:
            self.get_logger().error(f'OpenAI API call failed: {e}')
            return None
    
    def _call_openai_api_with_image(self, prompt: str, image_base64: str) -> Optional[str]:
        """Call OpenAI API with text and image prompt."""
        try:
            response = self.openai_client.chat.completions.create(
                model="gpt-4-vision-preview",
                messages=[
                    {
                        "role": "user",
                        "content": [
                            {"type": "text", "text": prompt},
                            {
                                "type": "image_url",
                                "image_url": {
                                    "url": f"data:image/jpeg;base64,{image_base64}"
                                }
                            }
                        ]
                    }
                ],
                max_tokens=150,
                temperature=0.7
            )
            
            return response.choices[0].message.content.strip()
        
        except Exception as e:
            self.get_logger().error(f'OpenAI API call with image failed: {e}')
            return None


def main(args=None):
    rclpy.init(args=args)
    
    llm_node = LLMNode()
    
    try:
        rclpy.spin(llm_node)
    except KeyboardInterrupt:
        pass
    finally:
        llm_node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/llm_node_cpp_basic/src/llm_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>
#include "liboai.h"
#include <memory>
#include <string>
#include <cstdlib>

class LLMNode : public rclcpp::Node {
public:
    LLMNode() : Node("llm_node_cpp_basic") {
        // Initialize OpenAI client
        openai_client_ = nullptr;
        setup_openai();
        
        // Create subscribers
        string_subscription_ = this->create_subscription<std_msgs::msg::String>(
            "/llm_input", 10,
            std::bind(&LLMNode::string_callback, this, std::placeholders::_1));
        
        // Create publisher for responses
        response_publisher_ = this->create_publisher<std_msgs::msg::String>(
            "/llm_response", 10);
        
        RCLCPP_INFO(this->get_logger(), "LLM Node (C++ Basic) initialized and listening for events...");
    }

private:
    void setup_openai() {
        const char* api_key = std::getenv("OPENAI_API_KEY");
        if (!api_key) {
            RCLCPP_ERROR(this->get_logger(), "OPENAI_API_KEY environment variable not set!");
            return;
        }
        
        try {
            openai_client_ = std::make_unique<liboai::OpenAI>();
            bool key_set = openai_client_->auth.SetKey(std::string(api_key));
            if (!key_set) {
                RCLCPP_ERROR(this->get_logger(), "Failed to set OpenAI API key");
                return;
            }
            RCLCPP_INFO(this->get_logger(), "OpenAI client initialized successfully");
        } catch (const std::exception& e) {
            RCLCPP_ERROR(this->get_logger(), "Failed to initialize OpenAI client: %s", e.what());
        }
    }
    
    void string_callback(const std_msgs::msg::String::SharedPtr msg) {
        if (!openai_client_) {
            RCLCPP_WARN(this->get_logger(), "OpenAI client not available, skipping request");
            return;
        }
        
        RCLCPP_INFO(this->get_logger(), "Received string message: %s", msg->data.c_str());
        
        try {
            // Create chat completion request
            liboai::Conversation convo;
            bool data_added = convo.AddUserData(msg->data);
            if (!data_added) {
                RCLCPP_ERROR(this->get_logger(), "Failed to add user data to conversation");
                return;
            }
            
               // Call OpenAI API
               auto response = openai_client_->ChatCompletion->create(
                   "gpt-5-nano",
                convo,
                std::nullopt,  // function_call
                std::nullopt,  // temperature (eg: 0.7f, use default for GPT-5-nano)
                std::nullopt,  // top_p
                std::nullopt,  // n
                std::nullopt,  // stream
                std::nullopt,  // stop
                std::nullopt,  // max_tokens (removed for GPT-5-nano compatibility)
                std::nullopt,  // presence_penalty
                std::nullopt,  // frequency_penalty
                std::nullopt,  // logit_bias
                std::nullopt   // user
               );
            
            // Update conversation with response
            bool updated = convo.Update(response);
            if (!updated) {
                RCLCPP_ERROR(this->get_logger(), "Failed to update conversation with response");
                return;
            }
            
            // Get the response text
            std::string response_text = convo.GetLastResponse();
            RCLCPP_INFO(this->get_logger(), "LLM Response: %s", response_text.c_str());
            
            // Publish response back to ROS
            auto response_msg = std_msgs::msg::String();
            response_msg.data = response_text;
            response_publisher_->publish(response_msg);
            
        } catch (const std::exception& e) {
            RCLCPP_ERROR(this->get_logger(), "Error processing string message: %s", e.what());
        }
    }
    
    std::unique_ptr<liboai::OpenAI> openai_client_;
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr string_subscription_;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr response_publisher_;
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    
    auto node = std::make_shared<LLMNode>();
    
    try {
        rclcpp::spin(node);
    } catch (const std::exception& e) {
        RCLCPP_ERROR(rclcpp::get_logger("main"), "Exception in main: %s", e.what());
    }
    
    rclcpp::shutdown();
    return 0;
}
